### PR TITLE
cmake: allow to build pytorch as a CMake subproject

### LINKED ISSUE #53980: Libtorch not buildable as a cmake "FetchContent" dependency in another project
## 🐛 Bug

Lines 330-331 in https://github.com/pytorch/pytorch/blob/master/caffe2/CMakeLists.txt explicitly set `TORCH_ROOT` and `TORCH_SRC_DIR` relative to `PROJECT_SOURCE_DIR`:
```
    set(TORCH_SRC_DIR "${PROJECT_SOURCE_DIR}/torch")
    set(TORCH_ROOT "${PROJECT_SOURCE_DIR}")
```

This breaks when including libtorch as a dependency in another project through `FetchContent`, and likely through `add_directory()`.  It should probably set these to `${CMAKE_CURRENT_SOURCE_DIR}/..` instead. 

## To Reproduce

1. Create an new empty project with a top-level CMakeLists.txt
1. Add pytorch as a dependency through FetchContent, so the resulting CMakeFile.txt looks something like:
```
cmake_minimum_required(VERSION 3.13 FATAL_ERROR)
project(foo CXX)
include(FetchContent)

FetchContent_Declare(
        pytorch
        GIT_REPOSITORY "https://github.com/pytorch/pytorch"
        GIT_TAG "release/1.8"
)
# Configure for minimal, static, openblas-enabled CPU library.
set(BUILD_SHARED_LIBS OFF)
set(USE_MKL OFF) # poor perf vs. openblas on older CPUs, ignores threading limits, cpuid-isms
set(USE_MKLDNN OFF) # ditto
set(USE_CUDA OFF)
set(USE_DISTRIBUTED OFF)
set(BUILD_PYTHON OFF)
set(USE_ROCM OFF)
set(USE_NCCL OFF)
set(BUILD_CAFFE2_OPS OFF)
set(BLAS_INFO "open")
set(WITH_BLAS "open")
set(USE_NUMA OFF)
FetchContent_MakeAvailable(pytorch)
```
1. mkdir build && cd build && cmake ..
Note this takes a while to checkout the git repo. 

1. Observe that the cmake command fails with:
```
--   USE_ZMQ               : OFF
--   USE_DISTRIBUTED       : OFF
--   USE_DEPLOY           : OFF
--   Public Dependencies  : Threads::Threads
--   Private Dependencies : pthreadpool;cpuinfo;qnnpack;pytorch_qnnpack;nnpack;XNNPACK;fbgemm;/usr/lib/x86_64-linux-gnu/libnuma.so;fp16;foxi_loader;rt;fmt::fmt-header-only;gcc_s;gcc;dl
-- Configuring done
CMake Error at build/_deps/pytorch-src/caffe2/CMakeLists.txt:719 (add_library):
  Cannot find source file:

    /home/john/git/test/torch/csrc/jit/frontend/error_report.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx


CMake Error at build/_deps/pytorch-src/caffe2/CMakeLists.txt:719 (add_library):
  No SOURCES given to target: torch_cpu


CMake Generate step failed.  Build files cannot be regenerated correctly.
```
Note that the source files are looking in the main project top level dir `/home/john/git/test` , not `/home/john/git/test/build/_deps/pytorch-src/`, the pytorch top level dir. 

There may be other issues as well but its the first one I've hit. 

## Expected behavior

I expect to be able to include libtorch as a dependency in in another CMake project. 
Current workaround is to build in a completely separate repo, install, and build from there. 

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): release/1.8 branch
 - OS (e.g., Linux): Linux 
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): see above
 - Python version: 3.8
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information:

## Additional context

I'm excited to test out static linking now that that's an option. I run MLC@Home, a volunteer distributed computing project running a client program on thousands of computers (x86, arm, windows, linux, etc..) to train thousands of neural networks for research. The client uses PyTorch (c++ bindings).  Having users download both a client and all the dependent libs has been a pain point since the beginning. It would be great if we could statically link our client into a single binary. 

cc @malfet @seemethere @walterddr @yf225 @glaringlee

### PR DESCRIPTION
**Pull request by SpaceIm**
Fixes https://github.com/pytorch/pytorch/issues/53980

Maybe it would be nice to find why some files are generated in `CMAKE_BINARY_DIR` instead of `CMAKE_CURRENT_BINARY_DIR` or `Torch_BINARY_DIR` or `PROJECT_BINARY_DIR`, but there is a lot of indirection in the logic of pytorch build files, so I was not able to find where it comes from.

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -614,7 +614,7 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       set(DELAY_LOAD_FLAGS "")
     endif()
     target_link_libraries(caffe2_nvrtc ${CUDA_NVRTC} ${CUDA_CUDA_LIB} ${CUDA_NVRTC_LIB} ${DELAY_LOAD_FLAGS})
-    target_include_directories(caffe2_nvrtc PRIVATE ${CUDA_INCLUDE_DIRS})
+    target_include_directories(caffe2_nvrtc PRIVATE ${CUDA_INCLUDE_DIRS} ${CMAKE_BINARY_DIR})
```
Comment (malfet) [inline caffe2/CMakeLists.txt:617]: Can you please explain why this change is necessary?

**Review by malfet — APPROVED** [review]

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (SpaceIm) [reply caffe2/CMakeLists.txt:617]: Because this target compiles `ATenNVRTC.cpp`, which includes `ATenNVRTC.h` which includes `ATen/cuda/ATenCUDAGeneral.h` which includes... `c10/macros/Export.h`, generated in `${CMAKE_BINARY_DIR}`.

Actually, I believe that `caffe2_nvrtc` target in ROCm branch also requires this fix.

**Comment by SpaceIm** [comment]
> thanks for the contribution. could you add test to verify this PR fixes #53980?
> you can reference how custom backend is being tested in: https://github.com/pytorch/pytorch/tree/master/test/custom_backend

I'll try to add a test.

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (SpaceIm) [reply caffe2/CMakeLists.txt:617]: I'm wondering if `c10` shouldn't be added to dependencies of `caffe2_nvrtc` through `target_link_libraries()` instead of this `target_include_directories()`, it would make more sense, specially in ROCm branch where several `c10` headers are included.

**Comment by bsergean** [comment]
The 2 bots that failed are

ci/circleci: pytorch_linux_xenial_py3_6_gcc5_4_build 
ci/circleci: pytorch_xla_linux_bionic_py3_6_clang9

Are those using old version of cmake maybe, which don't support what's required in this patch. Maybe those builders aren't used anymore, and re-running it today would work.
